functional_programming_tests/test/with_parameters.cpp

```diff
@@ -0,0 +1,484 @@
+#include <chrono>
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <example_srvs/srv/get_path.hpp>
+#include <example_srvs/srv/set_map.hpp>
+#include <gtest/gtest.h>
+#include <std_msgs/msg/u_int8_multi_array.hpp>
+
+using namespace std::chrono_literals;
+
+// TODO: Add more comments
+
+// production_code.h/cc
+// X, Y position
+struct Position {
+  size_t x;
+  size_t y;
+};
+
+template <typename T>
+class Map {
+ public:
+  Map(){};
+  Map(std::vector<std::vector<T>> data) : data_{data} {};
+  T at(Position const& pos) const { return data_.at(pos.y).at(pos.x); }
+  std::vector<std::vector<T>>& get_data() { return data_; }
+  std::vector<std::vector<T>> const& get_data() const { return data_; }
+  std::pair<int, int> dim() const {
+    return std::make_pair(data_.size(), data_.at(0).size());
+  }
+
+ private:
+  std::vector<std::vector<T>> data_;
+};
+
+using Path = std::vector<Position>;
+
+// Operator overload to print path
+std::ostream& operator<<(std::ostream& os, std::vector<Position> const& path) {
+  for (auto const& pose : path) {
+    os << "(" << pose.x << ", " << pose.y << ")\n";
+  }
+  return os;
+}
+
+// Operator overload for position comparison
+bool operator==(Position const& lhs, Position const& rhs) {
+  return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+template <typename T>
+struct CollisionChecker {
+  /**
+   * @brief Checks if the robot is colliding with an obstacle or out of bounds
+   * @param robot_size The size of the robot in pixels
+   */
+  CollisionChecker(int robot_size) : robot_size_{robot_size} {}
+
+  /**
+   * @brief Checks if the robot is colliding with an obstacle or out of bounds
+   * @param map The map to check for collision
+   * @param x position of the center of robot
+   * @param y position of the center of robot
+   * TODO: This check doesn't make any physical sense. The robot should be
```

Comment:
While that is true I think it can just be a comment for the moment where hey, this doesn't make physical sense but its here because it's simple right now.

Reply:
yeah. for now. we'll see where we get